### PR TITLE
Updating GitHub templates (contributing, issue, pull request)

### PR DESCRIPTION
**Pull request by syclik**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`: N/A
- [x] Run cpplint: `make cpplint`: N/A
- [x] Declare copyright holder and open-source license: see below

#### Summary:
Overhauling the GitHub templates.

#### Intended Effect:
Make it clearer what the pull request should include.

#### How to Verify:
This is all doc. Just read.

#### Side Effects:
This is going to affect all new pull requests.

#### Documentation:
All doc.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Daniel Lee

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by syclik** [comment]
@seantalts, mind taking a look? I think it can be improved a lot, but this is a step up from where we were. What do you think? Feel free to change at will. I'm trying to make it clearer what's expected, but maybe I missed the mark.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -1,23 +1,56 @@
...
 
-#### Documentation:
+    - all unit tests pass
+    - continuous integration passes
 
-#### Copyright and Licensing
+- [ ] the changes are maintainable
 
-Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):
+    - the code design is standard C++ or there's a good reason it's not
```
Comment (seantalts) [inline .github/PULL_REQUEST_TEMPLATE.md]: What does this mean, standard C++? Maybe link to our guidelines wiki (the one where we say how we're different from Google) or something? Every company uses a different subset of C++ from what I can tell.

**Comment by syclik** [comment]
Thanks! Any other changes? I was hoping you’d pick up on something I
missed. (Incremental improvement is of course ok, so if something comes up
later, just fix it then.)

Standard C++: sorry I was really vague. I don’t know how to explain what
I’m trying to say well. What I’d want is C++ that looks like C++. Some of
our challenge is that our contributions often come from people that are
more fluent in R or Python. Some of the code just doesn’t look like
idiomatic C++ when it’s more of a port; it’s something that auto formatting
wouldn’t really pick up on. Does that make sense? I’m happy to link to our
code quality link if that’s sufficient. I guess the expectation is if a
reviewer says it doesn’t look like C++ and can provide solid suggestions to
make it look more stylistically like C++, the submitter should try to
comply.

Let me think about how to make that clearer. If you think of something,
please replace the text.

**Comment by bob-carpenter** [comment]
I think "idiomatic" is the usual term of art for the concept you're trying to get at.

How do you feel about enforcing good C++11 practices?  Is that part of being idiomatic these days?  If so, our current code is a poor example of best practices.

**Comment by seantalts** [comment]
I was under the impression that there wasn't really such a thing as universally idiomatic C++, though I think maybe there is a push from a few C++ heavyweights to create such a thing in the past couple of years. I may just not know enough about the community.

**Comment by syclik** [comment]
Yes, there’s no universally idiomatic C++. But some code is just doesn’t
look or feel like C++, let alone look like the rest of our code base.

What I’m trying to do is to set the expectation that the contributors
should be using idiomatic C++ (what we consider idiomatic) and that the
reviewer can ask for changes based on that. Yes, this is subjective.
Ideally when this sort of comment is made, it’s clear to the contributor
that there’s an alternative way of implementing the same thing that makes
sense to other Math developers.

I think at the heart, that’s what I’m trying to get at. If I read some code
and have to spend an hour to figure out an implementation due to a
non-standard design pattern or some wacky recursive implementation or if
statements nested 20 deep and there was a way to implement it so that
someone that knows C++ could figure it out in a fraction of the time, I
think that’s absolutely fair to request.

**Comment by seantalts** [comment]
Ahh, yes, that makes complete sense. I will try to think up a way to pithily describe that... Maybe we just say something like "(our version of) idiomatic" and leave it vague. It's not even mirroring the code around it, because we have aspirations to update much of that for C++11...

**Comment by drezap** [comment]
If there was a link to a style guide that we can follow (when possible), I'd be happy to follow that to make sure stuff is more  "idiomatic", (i.e. should I be traversing by pointer, should I be writing a reference instead of indexing the matrix).

**Comment by bob-carpenter** [comment]
For layout, we autoformat. 

No, there's not a style guide.  We are looking for standard, contemporary C++ to the extent supported by our compiler restrictions.  None of us are really high-end C++ specialists, so this is all a work in progress.

No, you shouldn't be doing anything by pointer unless you're down in the arena-memory management aspect of writing autodiff.  You shouldn't ever need to malloc/new or delete/free---contemporary C++ style largely encapsulates that to the extent possible in the RAII pattern.  Algorithms, iterators, and lambdas are all your friend, as is static branchpoint detection either via conditionals or template specializations.

As an intro for style, I'd recommend Scott Meyers's latest Effective C++ and the C++11/14 supplement.  I liked Herb Sutter's blog posts on C++11, but his books haven't been updated.  

For templates, I recommend Vandevoorde an Josuttis (second edition for C++11).  The first edition was my favorite C++ book.  Alexandrescu's book is good, and more aimed at programmers than computer scientists, but it hasn't been updated for C++11.

I'd also highly recommend reading Agner Fogg's web resources for optimization if you are going to be trying to write performant code.

The easiest thing to do is catch up on the background reading (also lots of good blogs, but then you need to know what to look for---Sutters' series on C++11 was good), and then give it a shot.  We can correct during code review or before.

**Comment by drezap** [comment]
@bob-carpenter very helpful, thank you.. not the place for this but which C++ blogs are you following currently or have in the past?

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (syclik) [reply .github/PULL_REQUEST_TEMPLATE.md]: I'll change that to "idiomatic" before we merge. We can make adjustments to the pull request template if it seems too long / cumbersome.

**Pull request merged**